teachertool/src/components/MakecodeFrame.tsx

```diff
@@ -1,140 +1,31 @@
 /// <reference path="../../../built/pxteditor.d.ts" />
 import * as React from "react";
-import { useState, useEffect, useRef } from "react";
-import { ProgressBar } from "react-common/components/controls/ProgressBar";
+import { useEffect, useRef } from "react";
+import { getIframeRef, removeIframeRef } from "../services/makecodeEditorService";
 interface MakeCodeFrameProps {
     pageSourceUrl: string;
-    highContrast?: boolean;
-    onFrameOpen: () => void;
-    onFrameClose: () => void;
     tutorialEventHandler?: (event: pxt.editor.EditorMessageTutorialEventRequest) => void;
 }
 
-type FrameState = "loading" | "no-project" | "opening-project" | "project-open" | "closing-project";
-
-interface PendingMessage {
-    original: pxt.editor.EditorMessageRequest;
-    handler: (response: any) => void;
-}
-
 export const MakeCodeFrame: React.FC<MakeCodeFrameProps> =
-    ( { pageSourceUrl,
-        highContrast,
-        onFrameClose,
-        onFrameOpen
-    } ) => {
-    let ref: HTMLIFrameElement | undefined;
-    const messageQueue: pxt.editor.EditorMessageRequest[] = [];
-    let nextId: number = 0;
-    let pendingMessages: {[index: string]: PendingMessage} = {};
-    const [frameState, setFrameState] = useState<FrameState>("loading");
-    const [loadPercent, setLoadPercent] = useState(0);
-    const [workspaceReady, setWorkspaceReady] = useState(false);
+    ( { pageSourceUrl} ) => {
 
     let frameRef = useRef<HTMLIFrameElement>(null);
```

Comment:
I have mixed feelings on this. On the one hand, passing the ref does keep all the logic very contained within the service, but hanging onto a ref like that feels less safe to me than passing a callback to send a message to the iframe instead. The function passed would essentially just be something like:

```ts
function sendMessageToIFrame(message) {
   frameRef!.contentWindow!.postMessage(message, "*");`)
}
```

Then the service calls into that instead of calling `makecodeEditorRef!.contentWindow!.postMessage(message, "*");`

Reply:
Hmm, interesting. So you would suggest to have a function in the service of, let's say, `connectToFrame(func: () => void)`, save that passed-in function to something inside the service, and then call that function when we need to send a message to the iframe. Am I following that correctly?

I kinda like this infrastructure. However, following this can get kinda convoluted. Because if we don't have the ref somewhere in the service, then we would need to have the adding and removing event listeners in the `MakeCodeFrame`. This is fine, but that would then mean that we also need to expose `onMessageReceived` from the service for the iframe to use, if we want to keep the `onMessageReceived` logic in the service. So, in the iframe, you would be sending, to the service, a way for the service to communicate to the iframe and receiving, from the service, a function for what to do when messages are received. 

I think this infrastructure could work, but I'm not a huge fan of the circular logic that's involved. I like just having one entry point for the service and iframe to communicate.

Reply:
It seems like the event listener is actually on the window, not the iframe ref, so I'm wondering if it's actually necessary to have the ref for that. If we care about the iframe loaded event, then you're right, the component would have to bubble that up in addition to the "sendMessageToIFrame" function it sends (or it could only send that callback when the iframe is ready).

I don't think it makes a huge difference either way, as long as we're careful to be resilient against the ref changing. One of Eric's comments suggests another approach to doing that too, which might also work.

Reply:
Oh, that's a good point. I feel like a combination of your suggestion and Eric's suggestion would be really solid, but I might be misunderstanding something, and TypeScript might also not like the fact that the "ref" variable wouldn't be used. Let me know what you think.

Basically, we have the `handleIframeRef` function that Eric suggested and, instead of setting the ref for the service in the body of that function, we send the message function that you suggested through to the service. This way, we have a really solid way of knowing that the iframe is initialized and we don't have to pass a ref to the service.

Reply:
I think if you're going with the handleIframeRef approach, you can just pass the ref around. Passing the function instead of the ref would probably be a little redundant at that point...but Eric's approach may be better in this case, since the service would have a clearer sense for when the iframe gets cleared for whatever reason.